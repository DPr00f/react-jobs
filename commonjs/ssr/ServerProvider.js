'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ClientProvider = require('./ClientProvider');

var _ClientProvider2 = _interopRequireDefault(_ClientProvider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } // eslint-disable-next-line import/no-extraneous-dependencies


var ServerProvider = function (_Component) {
  _inherits(ServerProvider, _Component);

  function ServerProvider(props) {
    _classCallCheck(this, ServerProvider);

    var _this = _possibleConstructorReturn(this, (ServerProvider.__proto__ || Object.getPrototypeOf(ServerProvider)).call(this, props));

    if (props.runJobsExecContext) {
      _this.jobStates = props.runJobsExecContext.getState().jobsState;
    } else {
      _this.jobStates = {};
    }
    return _this;
  }

  _createClass(ServerProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var _this2 = this;

      var context = {
        reactJobsServer: {
          registerJobState: function registerJobState(jobID, jobState) {
            _this2.jobStates[jobID] = jobState;
            if (_this2.context.runJobsExecContext) {
              _this2.context.runJobsExecContext.registerJobState(jobID, jobState);
            }
          },
          getJobState: function getJobState(jobID) {
            return _this2.jobStates[jobID];
          }
        }
      };

      return context;
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _ClientProvider2.default,
        null,
        this.props.children
      );
    }
  }]);

  return ServerProvider;
}(_react.Component);

ServerProvider.contextTypes = {
  reactJobsRunner: _react.PropTypes.object
};
ServerProvider.childContextTypes = {
  reactJobsServer: _react.PropTypes.object.isRequired
};

exports.default = ServerProvider;