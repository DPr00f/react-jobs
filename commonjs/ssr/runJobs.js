'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = runJobs;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactTreeWalker = require('react-tree-walker');

var _reactTreeWalker2 = _interopRequireDefault(_reactTreeWalker);

var _utils = require('../utils');

var _ServerProvider = require('./ServerProvider');

var _ServerProvider2 = _interopRequireDefault(_ServerProvider);

var _createRunJobsExecContext = require('./createRunJobsExecContext');

var _createRunJobsExecContext2 = _interopRequireDefault(_createRunJobsExecContext);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getJobs(rootElement, rootContext) {
  var fetchRoot = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var jobs = [];

  (0, _reactTreeWalker2.default)(rootElement, function (element, instance, context) {
    var skipRoot = !fetchRoot && element === rootElement;
    if (instance && typeof instance.getExecutingJob === 'function' && !skipRoot) {
      var job = instance.getExecutingJob();
      if ((0, _utils.isPromise)(job)) {
        jobs.push({ job: job, element: element, context: context });

        // Tell walkTree to not recurse inside this component;  we will
        // wait for the query to execute before attempting it.
        return false;
      }
    }

    return undefined;
  }, rootContext);

  return jobs;
}

function runJobs(rootElement) {
  var rootContext = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var isRoot = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var processingElement = void 0;
  if (isRoot) {
    var runJobsExecContext = (0, _createRunJobsExecContext2.default)();
    rootContext.runJobsExecContext = runJobsExecContext; // eslint-disable-line no-param-reassign
    processingElement = _react2.default.createElement(
      _ServerProvider2.default,
      { runJobsExecContext: runJobsExecContext },
      rootElement
    );
  } else {
    processingElement = rootElement;
  }

  var resolveResult = function resolveResult() {
    return {
      appWithJobs: processingElement,
      state: rootContext.runJobsExecContext.getState(),
      STATE_IDENTIFIER: _constants.STATE_IDENTIFIER
    };
  };

  var runningJobs = getJobs(processingElement, rootContext, isRoot)
  // Map over the jobs and then...
  .map(function (_ref) {
    var job = _ref.job,
        element = _ref.element,
        context = _ref.context;
    return (
      // ... make sure they will continue executing jobs for their Children
      // when they are complete.
      job.then(function () {
        return runJobs(element, context,
        // We've just grabbed the job for element so don't try and get it again
        false);
      })
    );
  });

  return runningJobs.length > 0 ? Promise.all(runningJobs)
  // Swallow errors.
  .catch(function () {
    return undefined;
  }).then(function () {
    return resolveResult();
  }) : Promise.resolve(resolveResult());
}