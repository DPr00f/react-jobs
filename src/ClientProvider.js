/* @flow */

import { Children, Component, Element, PropTypes } from 'react';
import type { ClientProviderContext } from './types';

type React$Element = Element<*>;

type Props = {
  children?: React$Element,
  state?: Object,
};

let currentjobID = 0;

class ClientProvider extends Component {
  props: Props;

  getChildContext() {
    const context : ClientProviderContext = {
      reactJobsClient: {
        nextJobID: () => {
          currentjobID += 1;
          return currentjobID;
        },
      },
    };

    return context;
  }

  render() {
    return Children.only(this.props.children);
  }
}

ClientProvider.childContextTypes = {
  reactJobs: PropTypes.object.isRequired,
};

export default ClientProvider;
