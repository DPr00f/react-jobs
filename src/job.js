/* @flow */

import React, { Component, PropTypes } from 'react';
import { getDisplayName, isPromise } from './utils';
import type { ServerProviderContext } from './server/types';
import type { ClientProviderContext } from './types';

type Work = (props : Object) => any;

type State = {
  resolved?: boolean,
  inProgress: boolean,
  result?: any,
  error?: any,
  executingWork? : Promise<any>,
};

type Props = Object;

type ProviderContext = ServerProviderContext | ClientProviderContext;

export default function job(work : Work) {
  if (typeof work !== 'function') {
    throw new Error('You must provide a function to a react-jobs job declaration.');
  }

  // Oh I love you closures!
  let jobID = null;

  function withJobID(WrappedComponent) {
    const ComponentWithJobID = (props : Object, context : ProviderContext) => {
      if (!jobID) {
        jobID = context.reactJobs.nextJobID();
      }
      return <WrappedComponent {...props} jobID={jobID} />;
    };
    ComponentWithJobID.displayName = `${getDisplayName(WrappedComponent)}WithJobID`;
    ComponentWithJobID.contextTypes = { reactJobs: PropTypes.object };
    return ComponentWithJobID;
  }

  return function WrapComponentWithJob(WrappedComponent : Function) {
    class ComponentWithJob extends Component {
      props: Props;
      state: State;

      constructor(props : Props, context : ProviderContext) {
        super(props, context);
        if (context.reactJobsClient) {
          if (context.reactJobsClient.isJobResolved(this.props.jobID)) {
            // We are rehydrating on the client, therefore we will ignore this
            // render parse.
            return;
          }
        }
        if (context.reactJobsServer) {
          const jobResults = context.reactJobsServer.getJobResults(this.props.jobID);
          if (jobResults) {
            this.state = Object.assign(
              {},
              jobResults,
              { resolved: true },
            );
          }
        }
        if (!this.state) {
          this.state = { inProgress: false };
        }
      }

      componentWillMount() {
        if (this.state.resolved) {
          return;
        }

        const context : ProviderContext = this.context;
        const x = work(this.props);

        if (isPromise(x)) {
          const executingWork = x
            .then((result) => { this.setState({ inProgress: false, result }); return result; })
            .catch((error) => {
              console.warn('Job failed:\n', error); // eslint-disable-line no-console
              this.setState({ inProgress: false, error });
            })
            .then(() => {
              if (context.reactJobsServer) {
                context.reactJobsServer.registerJobResults(this.props.jobID, this.state);
              }
            });

          // Asynchronous result.
          this.setState({ inProgress: true, executingWork });
        } else {
          // Synchronous result.
          this.setState({ result: x });
        }
      }

      getExecutingJob() {
        if (!this.state.executingWork) {
          return undefined;
        }
        return this.state.executingWork;
      }

      render() {
        const { inProgress, result, error } = this.state;
        const jobProp = { inProgress, result, error };
        return <WrappedComponent {...this.props} job={jobProp} />;
      }
    }
    ComponentWithJob.displayName = `${getDisplayName(WrappedComponent)}WithJob`;
    ComponentWithJob.contextTypes = { reactJobs: PropTypes.object };
    return withJobID(ComponentWithJob);
  };
}
