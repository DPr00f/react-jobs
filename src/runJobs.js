/* @flow */

import { Children, Element } from 'react';
import { isPromise } from './utils';

type React$Element = Element<*>;
type Context = { [key: string]: any; };
type ElementVisitor =
  (element: React$Element, instance: ?Function, context: Context) => boolean | void;
type ElementJob = {
  job: Promise<any>,
  element: React$Element,
  context: Context,
}

// Recurse an React Element tree, running visitor on each element.
// If visitor returns `false`, don't call the element's render function
//   or recurse into its child elements
export function walkTree(
  element: React$Element,
  context: Context,
  visitor: ElementVisitor,
) {
  const Component = element.type;

  // a stateless functional component or a class
  if (typeof Component === 'function') {
    const props = Object.assign({}, Component.defaultProps, element.props);
    let childContext = context;
    let child;

    // Are we are a react class?
    // http://bit.ly/2j9Ifk3
    const isReactClassComponent = Component.prototype &&
      (Component.prototype.isReactComponent || Component.prototype.isPureReactComponent);

    if (isReactClassComponent) {
      const instance = new Component(props, context);
      // In case the user doesn't pass these to super in the constructor
      instance.props = instance.props || props;
      instance.context = instance.context || context;

      // Override setState to just change the state, not queue up an update.
      //   (we can't do the default React thing as we aren't mounted "properly"
      //   however, we don't need to re-render as well only support setState in
      //   componentWillMount, which happens *before* render).
      instance.setState = (newState) => {
        instance.state = Object.assign({}, instance.state, newState);
      };

      if (instance.componentWillMount) {
        instance.componentWillMount();
      }

      if (instance.getChildContext) {
        childContext = Object.assign({}, context, instance.getChildContext());
      }

      if (visitor(element, instance, context) === false) {
        return;
      }

      child = instance.render();
    } else { // just a stateless functional
      if (visitor(element, null, context) === false) {
        return;
      }

      child = Component(props, context);
    }

    if (child) {
      walkTree(child, childContext, visitor);
    }
  } else { // a basic string or dom element, just get children
    if (visitor(element, null, context) === false) {
      return;
    }

    if (element.props && element.props.children) {
      Children.forEach(element.props.children, (child: any) => {
        if (child) {
          walkTree(child, context, visitor);
        }
      });
    }
  }
}

function getJobs(
  rootElement : React$Element,
  rootContext : Context,
  fetchRoot : boolean = true,
) : ElementJob[] {
  const jobs = [];

  walkTree(rootElement, rootContext, (element, instance, context) => {
    const skipRoot = !fetchRoot && (element === rootElement);
    if (instance && typeof instance.getExecutingJob === 'function' && !skipRoot) {
      const job = instance.getExecutingJob();
      if (isPromise(job)) {
        jobs.push({ job, element, context });

        // Tell walkTree to not recurse inside this component;  we will
        // wait for the query to execute before attempting it.
        return false;
      }
    }

    return undefined;
  });

  return jobs;
}

export default function runJobs(
  rootElement : React$Element,
  rootContext : Context = {},
  fetchRoot : boolean = true,
) {
  const jobs = getJobs(rootElement, rootContext, fetchRoot);

  // no queries found, nothing to do
  if (!jobs.length) {
    return Promise.resolve(null);
  }

  // wait on each query that we found, re-rendering the subtree when it's done
  const mappedJobs = jobs.map(({ job, element, context }) =>
    // we've just grabbed the query for element so don't try and get it again
    job.then(() => runJobs(element, context, false)),
  );

  return Promise.all(mappedJobs).then(() => null);
}
