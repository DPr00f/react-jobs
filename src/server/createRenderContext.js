/* @flow */

import type { RenderContext } from '../sharedTypes';

export default function createRenderContext() : RenderContext {
  let currentjobID = 0;
  const resolvedJobs = {};
  return {
    setCurrentJobID: (jobID) => {
      currentjobID = jobID;
    },
    registerResolvedJob: (jobID) => {
      resolvedJobs[jobID] = true;
    },
    getState() {
      return {
        currentjobID,
        resolvedJobs,
      };
    },
  };
}
