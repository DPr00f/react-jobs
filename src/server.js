/* @flow */

import runJobs from './runJobs';

export default {
  runJobs,
};
