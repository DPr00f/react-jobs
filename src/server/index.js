/* @flow */

import createRenderContext from './createRenderContext';
import runJobs from './runJobs';
import ServerProvider from './ServerProvider';

export default {
  createRenderContext,
  runJobs,
  ServerProvider,
};
