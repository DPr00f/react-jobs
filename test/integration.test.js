/* @flow */

import React from 'react';
import { renderToString } from 'react-dom/server';
import { Foo, resolveAfter } from './__helpers__';
import ClientProvider from '../src/ClientProvider';
import ServerProvider from '../src/server/ServerProvider';
import job from '../src/job';
import runJobs from '../src/server/runJobs';
import createRenderContext from '../src/server/createRenderContext';
import type { RenderContext } from '../src/server/types';

const workTime = 10;

const createServerApp = (renderContext : RenderContext) => {
  const FooWithJob = job(() => resolveAfter(workTime, 'Hello world!'))(Foo);
  return (
    <ServerProvider renderContext={renderContext}>
      <FooWithJob />
    </ServerProvider>
  );
};

const createClientApp = (state : Object) => {
  const FooWithJob = job(() => resolveAfter(workTime, 'Hello world!'))(Foo);
  return (
    <ClientProvider state={state}>
      <FooWithJob />
    </ClientProvider>
  );
};

describe('integration', () => {
  it.only('works', () => {
    const serverRenderContext = createRenderContext();
    const serverApp = createServerApp(serverRenderContext);

    return runJobs(serverApp)
      .then(() => {
        // "Server" render
        const serverRender = renderToString(serverApp);
        expect(serverRender).toMatchSnapshot();
        const state = serverRenderContext.getState();

        // "Client" render
        const clientApp = createClientApp(state);
        const clientRender = renderToString(clientApp);
        expect(clientRender).toEqual(serverRender);
      });
  });
});
