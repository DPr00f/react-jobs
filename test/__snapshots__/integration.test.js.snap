exports[`integration SSR 1`] = `"<div data-reactroot="" data-reactid="1" data-react-checksum="1334186935">Hello world!</div>"`;

exports[`integration works 1`] = `"<div data-reactroot="" data-reactid="1" data-react-checksum="1984762928">In progress...</div>"`;
